legacy/signals/signals_test.go

```diff
@@ -0,0 +1,152 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package signals_test
+
+import (
+	"fmt"
+	"os"
+	"sync"
+	"syscall"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	"sigs.k8s.io/k8s-container-image-promoter/legacy/signals"
+)
+
+func TestLogSignal(t *testing.T) {
+	// Define what a test looks like.
+	type sigTest struct {
+		signal    os.Signal // Incoming signal to observe.
+		expected  []string  // Expected logs to be produced.
+		terminate bool      // Determine if signals.LogSignals() should return.
+	}
+	// Create multiple tests.
+	// NOTE: We are unable to observe SIGKILL or SIGSTOP, therefore we will not
+	// test with these syscalls.
+	sigTests := []sigTest{
+		{
+			signal: syscall.SIGIO,
+			expected: []string{
+				fmt.Sprintf("Encoutered signal: %s", syscall.SIGIO.String()),
+			},
+			terminate: false,
+		},
+		{
+			signal: syscall.SIGALRM,
+			expected: []string{
+				fmt.Sprintf("Encoutered signal: %s", syscall.SIGALRM.String()),
+			},
+			terminate: false,
+		},
+		{
+			signal: syscall.SIGALRM,
+			expected: []string{
+				fmt.Sprintf("Encoutered signal: %s", syscall.SIGALRM.String()),
+			},
+			terminate: false,
+		},
+		{
+			signal: syscall.SIGQUIT,
+			expected: []string{
+				fmt.Sprintf("Encoutered signal: %s", syscall.SIGQUIT.String()),
+				fmt.Sprintf("Exiting from signal: %s", syscall.SIGQUIT.String()),
+			},
+			terminate: true,
+		},
+		{
+			signal: syscall.SIGINT,
+			expected: []string{
+				fmt.Sprintf("Encoutered signal: %s", syscall.SIGINT.String()),
+				fmt.Sprintf("Exiting from signal: %s", syscall.SIGINT.String()),
+			},
+			terminate: true,
+		},
+		{
+			signal: syscall.SIGABRT,
+			expected: []string{
+				fmt.Sprintf("Encoutered signal: %s", syscall.SIGABRT.String()),
+				fmt.Sprintf("Exiting from signal: %s", syscall.SIGABRT.String()),
+			},
+			terminate: true,
+		},
+	}
+	// Capture all logs.
+	logs := []string{}
+	// Mock logrus.Debug statements.
+	signals.Debug = func(args ...interface{}) {
+		logs = append(logs, fmt.Sprint(args...))
+	}
+	// Determine if LogSignal invoked Stop().
+	terminated := func() bool {
+		return <-signals.ExitChannel
+	}
+	// Used for enforcing defaults before each test.
+	reset := func() {
+		logs = []string{}
+	}
+	// Run through all tests.
+	for _, st := range sigTests {
+		reset()
+		// Log the test signal.
+		signals.LogSignal(st.signal)
+		// Ensure the logs are correct.
+		require.EqualValues(t, st.expected, logs, "Unexpected signal observation logs.")
+		if st.terminate {
+			// Ensure Stop() was invoked if the test specifies.
+			require.True(t, terminated(), "LogSignal did not terminate on exit signal.")
+		}
+	}
+}
+
+func TestLogSignals(t *testing.T) {
```

Comment:
Nit: Add a comment somewhere in here describing what this test does. I think you're testing whether termination happens in situations where (1) we explicitly call signals.Stop(), or (2) where we pass in a signal that is recognized as one that should exit (line 147). Is this correct?

Reply:
Added comment.